### DISCUSSION TITLE
Setting the transaction isolation level

### DISCUSSION THREAD

**Discussion by ricogallo** [supabase/supabase#8798] [Questions] [answered]
Hi, is it possible to set the transaction isolation level instead doing something like this?

```plpgsql
-- this comes with drawbacks, such as not being able to COMMIT or ROLLBACK
create or replace procedure foo_proc()
set default_transaction_isolation = 'serializable'
as $$
begin
    insert into tab1 values (1);
end;
$$ language plpgsql;
```

In my particular case I need to use serializable transactions for a non-trivial use case. Better yet, is there a way to wrap any statements with a classic `BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE; <statements>; COMMIT;` transaction and set the isolation level outside?

**Comment by steve-chavez** [comment, marked as answer]
Hello,

What client are you using to connect to the db? Are you using the supabase-js library? 

> In my particular case I need to use serializable transactions for a non-trivial use case

Could you elaborate? Is it about handling concurrent updates?

**Reply by ricogallo** [reply]
> What client are you using to connect to the db? Are you using the supabase-js library?

Yes, this is the library I intend to use.

> Could you elaborate? Is it about handling concurrent updates?

It's about a matching engine for a stock-like exchange, where there are plenty of automatic concurrent reads that should be executed serially.

**Reply by steve-chavez** [reply]
I see. Right now all transactions use the `read committed` isolation.

> It's about a matching engine for a stock-like exchange, where there are plenty of automatic concurrent reads that should be executed serially.

It sounds that `repeatable read` might suit your use case then, see https://dba.stackexchange.com/questions/284744/postgres-repeatable-read-vs-serializable.

---

Regarding your question, right now there isn't a way to specify the isolation level for supabase-js. Support for pg 11 procedures are also pending - only functions allowed - but I assume this is not what you want since procedures can't return a result. 

We can add this, but we need to be careful with the interface because serializable isolation is costly, there should be a server-side whitelist on the tables that can use this isolation level.

**Reply by ricogallo** [reply]
Truth to be told, I might be ok with repeatable reads and avoiding lost updates. I am worried that I might incur in read/write dependency issues for transactions that I *might* need down the line. Adopting a Serializable isolation is a sort of guarantee (with performance drawbacks) that even in the worst case scenario the consistency is guaranteed.

Back to the point, it would be important for me to at least have a reapeatable read level and experiment with it and see if it works or not.

> It sounds that repeatable read might suit your use case then, see https://dba.stackexchange.com/questions/284744/postgres-repeatable-read-vs-serializable.

Thanks for the link.

**Reply by steve-chavez** [reply]
I definitely see the need, I've opened https://github.com/supabase/postgrest-js/issues/320

**Reply by ricogallo** [reply]
@steve-chavez I played around a bit with the SQL editor and I created the aforementioned procedure, and I actually found it under the Functions tab. Can I now assume that the limitations you mentioned affect only the supabase-js client?

**Reply by steve-chavez** [reply]
Yes, the supabase-js client right now can't invoke stored procedures(only functions).

**Reply by ricogallo** [reply]
Thanks for your help!

**Reply by ricogallo** [reply]
@steve-chavez actually I realized that the above code doesn't really set the transaction isolation level, so I can't rpc my way to this easily. Is there any escape hatch for this? I would consider a direct connection to the db but then I would lose the benefits of the auth module

**Reply by steve-chavez** [reply]
@ricogallo Hm. Maybe `dblink` could work, like on this [SO question](https://stackoverflow.com/questions/28299150/postgres-dblink-stored-procedure-call-with-specifying-transaction-level).

**Reply by ricogallo** [reply]
So I tried this approach, but I experienced some issues with `auth.uid()` and RLS in general, here's why:

```sql
-- create a test table
create table tab1 (name text);

-- a function to prove the uid is indeed read by the function
create or replace function my_function() returns void as $$
begin
  insert into tab1 (name) values (auth.uid());
end;
$$ language plpgsql;

-- function to invoke the dblink extension
create or replace function link_function() returns void as $$
begin
  PERFORM dblink_connect('host=localhost port=5432');
  PERFORM * from dblink('select my_function()') alias(col text);
end;
$$ language plpgsql;
```

Then I curl'd my way to test it:
```bash
# this worked fine, it stored the auth.uid()
curl -X POST 'https://<proj>.supabase.co/rest/v1/rpc/my_func' \
-H "Content-Type: application/json" \
-H "apikey: SUPABASE_KEY" \
-H "Authorization: Bearer MY_JWT_TOKEN"

# this didn't save the auth.uid()
curl -X POST 'https://<proj>.supabase.co/rest/v1/rpc/link_function' \
-H "Content-Type: application/json" \
-H "apikey: SUPABASE_KEY" \
-H "Authorization: Bearer MY_JWT_TOKEN"

```

While I had some success invoking by rpc `my_function` directly and effectively storing the `auth.uid` into the table, I could not replicate the same behavior using the `link_function` above. In my opinion since dblink acts in a context of isolation from the main session, it's likely it doesn't share any session settings downstream.

**Reply by steve-chavez** [reply]
> In my opinion since dblink acts in a context of isolation from the main session, it's likely it doesn't share any session settings downstream.

Yes, but I think you could pass the auth context like:

```sql
declare
  auth_jwt_ctx = current_setting('request.jwt.claims', true);
begin
  PERFORM * from dblink('begin') alias(col text);
  PERFORM * from dblink(format($_$select set_config('request.jwt.claims', %s, true)$_$), auth_jwt_ctx) alias(col text);
  PERFORM * from dblink('select my_function()') alias(col text);
end;
```

[auth.uid is defined here](https://github.com/supabase/gotrue/blob/master/migrations/20220224000811_update_auth_functions.up.sql#L3-L12). Let me know if that works.

**Reply by ricogallo** [reply]
Ok we are almost there 😂.

This is the final version of the link function:
```sql
create or replace function link_function() returns void as $$
declare
  auth_jwt_ctx text;
begin
  SELECT * INTO auth_jwt_ctx from current_setting('request.jwt.claims', true);
  -- please note the host is localhost here, but to make it work properly I need both the password and the full db url
  PERFORM dblink_connect('host=localhost port=5432');
  PERFORM * from dblink('begin transaction isolation level serializable;') alias(col text);
  PERFORM * from dblink(format($_$select set_config('request.jwt.claims', '%s', true);$_$, auth_jwt_ctx)) alias(col text);
  PERFORM * from dblink('select my_function();') alias(col text);
  PERFORM * from dblink('commit;') alias(col text);
end;
$$ language plpgsql;
```

The function works fine if for an external authenticated user, storing the jwt, but it requires a connection string with a password to actually work, otherwise it returns an error.

```json
{
	"code": "2F003",
	"details": "Non-superusers must provide a password in the connection string.",
	"hint": null,
	"message": "password is required"
}
```
And when I actually add the password to the connection string, then I get this error instead
```json
{
	"code": "2F003",
	"details": "Non-superuser cannot connect if the server does not request a password.",
	"hint": "Target server's authentication method must be changed.",
	"message": "password is required"
}
```

I would like to avoid adding the db password and the db's url to the connection string, but I am not aware of any way to circumvent this.

**Reply by steve-chavez** [reply]
Sorry, I wasn't aware of that limitation :disappointed:

I think for now you could try a different postgres client and do a direct connection to the db.

The feature for tx isolation([ref](https://github.com/PostgREST/postgrest/issues/2468)) is not hard to implement but we'll start working on features again next week(doing QA/docs now).

**Reply by ricogallo** [reply]
That's absolutely fine, thanks for you help!
